pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/onmetal/gardener-extension-provider-onmetal/pkg/onmetal"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// bastionEndpoints collects the endpoints the bastion host provides; the
+// private endpoint is important for opening a port on the worker node
+// ingress network policy rule to allow SSH from that node, the public endpoint is where
+// the end user connects to establish the SSH connection.
+type bastionEndpoints struct {
+	private *corev1.LoadBalancerIngress
+	public  *corev1.LoadBalancerIngress
+}
+
+// Reconcile implements bastion.Actuator.
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	return a.reconcile(ctx, log, bastion, cluster)
+}
+
+func (a *actuator) reconcile(ctx context.Context, log logr.Logger, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	log.V(2).Info("Reconciling bastion host")
+	err := bastionConfigCheck(a.bastionConfig)
+	if err != nil {
+		return err
+	}
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to determine options: %w", err)
+	}
+
+	onmetalClient, namespace, err := onmetal.GetOnmetalClientAndNamespaceFromCloudProviderSecret(ctx, a.Client(), cluster.ObjectMeta.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get onmetal client and namespace from cloudprovider secret: %w", err)
+	}
+	infraStatus, err := getInfrastructureStatus(ctx, onmetalClient, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get infrastructure status: %w", err)
+	}
+
+	// TODO: Add NetworkPolicy related implementation
+
+	ignitionSecret, err := createIgnitionSecret(namespace, opt)
+	if err != nil {
+		return fmt.Errorf("failed to create ignition secret: %w", err)
+	}
+
+	bastionMachine := createMachine(namespace, a.bastionConfig, infraStatus, opt.BastionInstanceName, ignitionSecret.Name)
+
+	machine, err := applyMachine(ctx, onmetalClient, ignitionSecret, bastionMachine)
+	if err != nil {
+		return fmt.Errorf("failed to create machine: %w", err)
+	}
+
+	endpoints, err := getMachineEndpoints(machine)
+	if err != nil {
+		return fmt.Errorf("failed to get machine endpoints: %w", err)
+	}
+
+	if !endpoints.Ready() {
+		return &reconcilerutils.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        fmt.Errorf("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	log.V(2).Info("Successfully reconciled bastion host")
+	patch := client.MergeFrom(bastion.DeepCopy())
+	bastion.Status.Ingress = endpoints.public
```

Comment:
What about the private endoint?

Reply:
Doesn't that come later, when we add network policy implementation?